ceph/ceph-nvmeof
# tests: automated testing w/ github actions and state update component tests

# Relevant Changes

- Added event to trigger graceful stop to periodic updates thread on shutdown before forced stop. This is also useful for testing.
- Organized tests folder into subfolders for functional tests (rely on Ceph/SPDK) and component tests. This solves import issues for pytest when running in isolated environments.
- Added 16 state update tests to `tests/component/test_state.py`
- Github action to automate test runs

## Tests
For the following, "valid" keys refer to keys with one of the established prefix constants.
1. No change between local and omap
2. Invalid key addition
3. Multiple key additions of one valid prefix
4. Key additions of every valid prefix
5. Key removal of invalid prefix
6. Multiple key removals of one valid prefix
7. Key removals of every valid prefix
8. Value change on key of invalid prefix
9. Multiple value changes on keys of a single valid prefix
10. Value changes on keys of every valid prefix
11. Confirm reset of local state after state update
12. Confirm reset of local version after state update
13. Confirm lack of update when local version == omap_version
14. Confirm lack of update when local version > omap_version
15. Confirm periodic call for update
16. Confirm notify event interrupts periodic call for update
